packages/editor/src/components/autosave-monitor/index.js

```diff
@@ -1,73 +1,59 @@
 /**
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
-import { compose } from '@wordpress/compose';
+import { usePrevious, compose } from '@wordpress/compose';
+import { useRef, useEffect } from '@wordpress/element';
 import { withSelect, withDispatch } from '@wordpress/data';
+/**
+ * Internal dependencies
+ */
+import useScheduleSave from './use-scheduled-save';
 
-export class AutosaveMonitor extends Component {
-	componentDidUpdate( prevProps ) {
-		const {
-			isDirty,
-			editsReference,
-			isAutosaveable,
-			isAutosaving,
-		} = this.props;
+export function AutosaveMonitor( props ) {
+	const {
+		isDirty,
+		editsReference,
+		isAutosaveable,
+		isAutosaving,
+		interval,
+		autosave,
+	} = props;
+	const { scheduleSave, cancelSave } = useScheduleSave( interval, autosave );
 
-		// The edits reference is held for comparison to avoid scheduling an
-		// autosave if an edit has not been made since the last autosave
-		// completion. This is assigned when the autosave completes, and reset
-		// when an edit occurs.
-		//
-		// See: https://github.com/WordPress/gutenberg/issues/12318
+	// The edits reference is held for comparison to avoid scheduling an
+	// autosave if an edit has not been made since the last autosave
+	// completion. This is assigned when the autosave completes, and reset
+	// when an edit occurs.
+	//
+	// See: https://github.com/WordPress/gutenberg/issues/12318
+	const didAutosaveForEditsReference = useRef();
+	useEffect( () => {
+		didAutosaveForEditsReference.current = false;
+	}, [ editsReference ] );
 
-		if ( editsReference !== prevProps.editsReference ) {
-			this.didAutosaveForEditsReference = false;
-		}
+	const prevIsAutosaving = usePrevious( isAutosaving );
+	if ( ! isAutosaving && prevIsAutosaving ) {
+		didAutosaveForEditsReference.current = true;
+	}
 
-		if ( ! isAutosaving && prevProps.isAutosaving ) {
-			this.didAutosaveForEditsReference = true;
+	const firstRender = useRef( true );
+	useEffect( () => {
+		if ( firstRender.current ) {
+			firstRender.current = false;
+			return;
 		}
-
 		if (
-			prevProps.isDirty !== isDirty ||
-			prevProps.isAutosaveable !== isAutosaveable ||
-			prevProps.editsReference !== editsReference
+			isDirty &&
+			isAutosaveable &&
+			! didAutosaveForEditsReference.current
```

Comment:
It's very unclear to me why this check is necessary? Why not just always "schedule an autosave on each edit change"?

Reply:
I guess the reason is because you have "cancelSave" as an API to call explicitely while IMO, it should be handled internally in `useScheduleSave`. Basically if there's a new scheduled autosave  while one is awaiting to be triggered, cancel it and schedule the new one instead.

Reply:
I like the new approach.

Reply:
> Basically if there's a new scheduled autosave while one is awaiting to be triggered, cancel it and schedule the new one instead.

This would be the same as the current debounced approach which causes long periods without an autosave as long as the user is typing. As discussed in the issue, we don't need to limit the amount of these requests as the autosave would amend the last autosaved revision instead of creating a new one every time.

I want to make it so once an autosave is scheduled, no subsequent change would cancel it. It simply stays scheduled at a fixed point in time, and can only be cancelled if the AutosaveMonitor is told it shouldn't act (e.g. `isAutosaveable == false`).

I am struggling with the number of corner cases to handle though and have an idea for a much simpler and more bullet-proof approach - I will update this PR soon to show exactly what I mean.